### PR TITLE
PIM-7324: rework structure version provider

### PR DESCRIPTION
**Pull request by juliensnz**
**Description (for Contributor and Core Developer)**

Currently, we use the versioning table to provide a catalog structure version. Each time you update a family, an attribute, etc, the structure version is updated (last update timestamp). The problem with this solution is that you rely on the versioning table: it can get pretty big (over millions of lines in some cases).
The idea of this fix is to replace the usage of the versioning table by a dedicated table with only the resource name and the last update date.

Todo:
- [x] add tests
- [x] add migration
- [ ] update changelog

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/CreateTable.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        ];
+    }
+
+    /**
+     * Add the csv format
+     */
+    public function onPostDBCreate()
+    {
+        $sql = <<<'SQL'
+DROP TABLE IF EXISTS `akeneo_structure_version_last_update`;
```
Comment (ahocquard) [inline src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/CreateTable.php]: @LaurentPetard without using doctrine ;)

File: src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/UpdateStructureVersion.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        return [
+            StorageEvents::POST_SAVE => 'onPostDBCreate'
+        ];
+    }
+
+    /**
+     * Add the csv format
+     */
+    public function onPostDBCreate(GenericEvent $event)
```
Comment (ahocquard) [inline src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/UpdateStructureVersion.php]: phpdoc + function name

File: src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/UpdateStructureVersion.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        $sql = <<<'SQL'
+REPLACE INTO akeneo_structure_version_last_update SET resource_name = :resource_name, logged_at = :logged_at;
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/UpdateStructureVersion.php]: isn't a where missing?

File: src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/CreateTable.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Listener on the install command to create the structure version table
+ *
+ * @author    Julien Sanchez <julien@akeneo.com>
+ * @copyright 2018 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class CreateTable implements EventSubscriberInterface
```
Comment (LaurentPetard) [inline src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/CreateTable.php]: Why a Subscriber instead of a Listener ?

Comment (juliensnz) [reply src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/CreateTable.php]: What would be the difference?

File: src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/UpdateStructureVersion.php (hunk 3 above)
Comment (juliensnz) [reply src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/UpdateStructureVersion.php]: No where clause in a replace into: https://stackoverflow.com/questions/3003244/does-replace-into-have-a-where-clause

File: src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/UpdateStructureVersion.php (hunk 2 above)
Comment (juliensnz) [reply src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/UpdateStructureVersion.php]: What do you want me to change in the function name?

File: src/Pim/Bundle/EnrichBundle/Provider/StructureVersion/StructureVersionProvider.php [hunk 5]
```diff
@@ -1,55 +0,0 @@
-<?php
```
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/Provider/StructureVersion/StructureVersionProvider.php]: CHANGELOG

File: src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/TableCreator.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Pim\Bundle\EnrichBundle\StructureVersion\EventListener;
+
+use Pim\Bundle\InstallerBundle\Event\InstallerEvents;
+use Symfony\Bridge\Doctrine\RegistryInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Listener on the install command to create the structure version table
```
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/TableCreator.php]: It miss a migration for 2.2 2.3, no?
Plus, you drop the previous information in the database without transferring it into the new ones, is it acceptable ?

Comment (juliensnz) [reply src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/TableCreator.php]: This PR is in 2.3 :) But you are right, a migration is missing.

Comment (juliensnz) [reply src/Pim/Bundle/EnrichBundle/StructureVersion/EventListener/TableCreator.php]: previously we were using the versioning table. I think we should not drop the versioning table ;)

File: src/Pim/Bundle/EnrichBundle/Resources/config/structure_version.yml [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        class: '%pim_enrich.provider.structure_version.class%'
+        arguments:
+            - '@doctrine'
+        calls:
+            - [addResource, ['%pim_catalog.entity.locale.class%']]
+
+    pim_structure_version.provider.structure_version.family:
+        class: '%pim_enrich.provider.structure_version.class%'
+        arguments:
+            - '@doctrine'
```
Comment (ahocquard) [inline src/Pim/Bundle/EnrichBundle/Resources/config/structure_version.yml]: better to inject directly the connection (for coupling)

File: src/Pim/Bundle/EnrichBundle/StructureVersion/Provider/StructureVersion.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        $this->doctrine = $doctrine;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getStructureVersion()
+    {
+        $sql = <<<'SQL'
+SELECT last_update
```
Comment (ahocquard) [inline src/Pim/Bundle/EnrichBundle/StructureVersion/Provider/StructureVersion.php]: `select MAX()` seems to be more appropriated for what you expect.

File: tests/legacy/features/product/invalidate_cache.feature [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  I need to be able to see the last structure data
+
+  Background:
+    Given a "footwear" catalog configuration
+    And I am logged in as "Julia"
+    And the following product:
+      | sku   | family |
+      | boots | boots  |
+
+  Scenario: Successfully clear the cache when updating a family label
```
Comment (ahocquard) [inline tests/legacy/features/product/invalidate_cache.feature]: I would not write it this way and create a dedicated context.

Given a family with label "Boots"
When the user "XXX" modify the family label with "Awesome boots"
Then I see the up-to-date label of this family in the product page

File: features/product/invalidate_cache.feature [hunk 10]
```diff
@@ -0,0 +1,17 @@
+@javascript
```
Comment (jjanvier) [inline features/product/invalidate_cache.feature]: wrong folder

**Pull request merged**